### DISCUSSION TITLE
Validating that something is undefined or a string that matches a regex

### DISCUSSION THREAD

**Discussion by gregmarr** [colinhacks/zod#2748] [Q&A] [answered]
This succeeds when given an input of `{}`.

```ts
z.object( {
    version: z.undefined()
        .or( z.string() )
        .optional()
} )
```

But this fails because there isn't a version field that matches the regex:

```ts
z.object( {
    version: z.undefined()
        .or( z.string().regex( /^v\d+\.\d+\.\d+$/ ) )
        .optional()
} )
```

Why is the regex validation in the or case triggering when it's already passed the `.undefined()` case?

**Comment by JacobWeisenburger** [comment, marked as answer]
I don't fully understand your question. Everthing seems to be working as expected.

```ts
const schema = z.object( {
    version: z.undefined().or( z.string().regex( /^v\d+\.\d+\.\d+$/ ) ).optional()
} )

console.log( schema.parse( {} ) ) // {}
console.log( schema.parse( { version: 'v1.2.3' } ) ) // { version: 'v1.2.3' }
```

Please let me know if I'm missing something.

Also, `z.undefined().or( z.string().regex( /^v\d+\.\d+\.\d+$/ ) ).optional()` is the same as `z.string().regex( /^v\d+\.\d+\.\d+$/ ).optional()`

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
